File: 2022/day04.cpp
// https://adventofcode.com/2022/day/4

#include "../common/common.h"
#include "../common/rust.h"

#include <array>
#include <print>
#include <ranges>
#include <span>
#include <string_view>

struct range : public aoc::closed_range<int> {
  using base_t = aoc::closed_range<int>;
  constexpr range(std::array<int, 2> range_array)
      : base_t{range_array[0], range_array[1]} {}
};

using Input = Vec<std::pair<range, range>>;

auto parse(String const& filename) -> Input {
  return aoc::views::read_lines(filename) |
         stdv::transform([](std::string_view line) {
           auto [firstElfStr, secondElfStr] = aoc::split_once(line, ',');
           return std::pair{range{aoc::split_once<int>(firstElfStr, '-')},
                            range{aoc::split_once<int>(secondElfStr, '-')}};
         }) |
         aoc::ranges::to<Input>();
}

template <bool use_overlaps>
fn solve_case(std::span<const std::pair<range, range>> input) -> u32 {
  u32 score = 0;
  for (let& [ firstElf, secondElf ] : input) {
    if constexpr (use_overlaps) {
      score += static_cast<u32>(firstElf.overlaps_with(secondElf));
    } else {
      score += static_cast<u32>(firstElf.contains(secondElf) ||
                                secondElf.contains(firstElf));
    }
  }
  return score;
}

int main() {
  std::println("Part 1");
  let example = parse("day04.example");
  AOC_EXPECT_RESULT(2, solve_case<false>(example));
  let input = parse("day04.input");
  AOC_EXPECT_RESULT(582, solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(4, solve_case<true>(example));
  AOC_EXPECT_RESULT(893, solve_case<true>(input));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day04.rs
struct Range {
    lo: i32,
    hi: i32,
}
impl Range {
    fn contains(&self, other: &Range) -> bool {
        self.lo <= other.lo && other.hi <= self.hi
    }
    fn overlaps_with(&self, other: &Range) -> bool {
        self.lo <= other.hi && other.lo <= self.hi
    }
}

fn parse_range(s: &str) -> Range {
    let (lo, hi) = s.split_once('-').unwrap();
    Range {
        lo: lo.parse().unwrap(),
        hi: hi.parse().unwrap(),
    }
}

type Input = Vec<(Range, Range)>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
        .into_iter()
        .map(|line| {
            let (a, b) = line.split_once(',').unwrap();
            (parse_range(a), parse_range(b))
        })
        .collect()
}

fn solve_case1(input: &Input) -> usize {
    input
        .iter()
        .filter(|(a, b)| a.contains(b) || b.contains(a))
        .count()
}

fn solve_case2(input: &Input) -> usize {
    input.iter().filter(|(a, b)| a.overlaps_with(b)).count()
}

fn main() {
    println!("Part 1");
    let example = parse("day04.example");
    assert_eq!(2, solve_case1(&example));
    let input = parse("day04.input");
    assert_eq!(582, solve_case1(&input));

    println!("Part 2");
    assert_eq!(4, solve_case2(&example));
    assert_eq!(893, solve_case2(&input));
}
